testhelpers/org.eclipse.n4js.ui.tests.helper/src/org/eclipse/n4js/tests/builder/AbstractBuilderTest.java

```diff
@@ -211,6 +211,25 @@ public void waitForAutoBuild(boolean assertValidityOfXtextIndex) {
 			assertXtextIndexIsValid();
 	}
 
+	/**
+	 * Trigger a blocking, incremental build. This is usually to be favored over {@link #waitForAutoBuild()} since it is
+	 * a mere method call and does not involve job joining or similar troublesome stuff.
+	 */
+	public void waitForIncrementalBuild() {
+		waitForIncrementalBuild(true);
+	}
+
+	/**
+	 * Wait for the incremental build and optionally assert the Xtext index to be valid.
+	 *
+	 * @see #waitForIncrementalBuild()
+	 */
+	public void waitForIncrementalBuild(boolean assertValidityOfXtextIndex) {
+		IResourcesSetupUtil.waitForBuild();
+		if (assertValidityOfXtextIndex)
+			assertXtextIndexIsValid();
+	}
```

Comment:
We have the following set of _wait_ methods:
* waitForAutoBuild()
* waitForAutoBuild(boolean)
* waitForNotReallyBuildButHousekeepingJobs()
* waitForIncrementalBuild()
* waitForIncrementalBuild(boolean)

When should we use `waitForAutoBuild` instead of `waitForIncrementalBuild`? How about `waitForNotReallyBuildButHousekeepingJobs`?

Reply:
In general: yes. I'd rather like to remove all the stuff that deals with the async AutoBuildJob and replace it by the blocking and immediate call to waitForIncrementalBuild. This would greatly speed up the test execution and simplify the reasoning about test failures.
A few dedicated AutoBuildJob tests could be used in addition to ensure that the autobuild is correctly triggered, but that is more of a paranoid testing of the builder infrastructure.